File: src/api/events.ts
import { listen } from "@tauri-apps/api/event";

export interface RecordingLevels {
  sys: number;
  mic: number;
}

export interface RecordingStatePayload {
  recording: boolean;
  duration: number;
  sessionId: string | null;
}

export interface TranscribeProgress {
  sessionId: string;
  phase: string;
  fraction: number | null;
}

export interface TranscribeDone {
  sessionId: string;
  error: string | null;
}

export interface ModelProgress {
  modelId: string;
  fraction: number;
  error: string | null;
  done: boolean;
}

export interface MinutesDone {
  sessionId: string;
  file: string | null;
  error: string | null;
}

export const events = {
  onRecordingLevels: (cb: (p: RecordingLevels) => void) =>
    listen<RecordingLevels>("recording://levels", (e) => cb(e.payload)),
  onRecordingState: (cb: (p: RecordingStatePayload) => void) =>
    listen<RecordingStatePayload>("recording://state", (e) => cb(e.payload)),
  onRecordingNoInput: (cb: (p: { sys: boolean; mic: boolean }) => void) =>
    listen<{ sys: boolean; mic: boolean }>("recording://no-input", (e) => cb(e.payload)),
  onRecordingAutoStopped: (cb: (p: { reason: string; minutes: number }) => void) =>
    listen<{ reason: string; minutes: number }>("recording://auto-stopped", (e) =>
      cb(e.payload),
    ),
  onLiveTranscript: (
    cb: (p: { source: string; finalized: string; volatile: string }) => void,
  ) =>
    listen<{ source: string; finalized: string; volatile: string }>(
      "live://transcript",
      (e) => cb(e.payload),
    ),
  onLiveTranslation: (cb: (p: { source: string; text: string }) => void) =>
    listen<{ source: string; text: string }>("live://translation", (e) => cb(e.payload)),
  onSessionsChanged: (cb: () => void) => listen("sessions://changed", () => cb()),
  onTranscribeProgress: (cb: (p: TranscribeProgress) => void) =>
    listen<TranscribeProgress>("transcribe://progress", (e) => cb(e.payload)),
  onTranscribeDone: (cb: (p: TranscribeDone) => void) =>
    listen<TranscribeDone>("transcribe://done", (e) => cb(e.payload)),
  onModelProgress: (cb: (p: ModelProgress) => void) =>
    listen<ModelProgress>("model://progress", (e) => cb(e.payload)),
  onMinutesDone: (cb: (p: MinutesDone) => void) =>
    listen<MinutesDone>("minutes://done", (e) => cb(e.payload)),
  onPlayerPosition: (cb: (p: import("./commands").PlayerPosition) => void) =>
    listen<import("./commands").PlayerPosition>("player://position", (e) => cb(e.payload)),
};

File: src-tauri/src/events.rs
use serde::Serialize;
use tauri::{AppHandle, Emitter};

pub const RECORDING_LEVELS: &str = "recording://levels";
pub const RECORDING_STATE: &str = "recording://state";
pub const RECORDING_NO_INPUT: &str = "recording://no-input";
pub const RECORDING_AUTO_STOPPED: &str = "recording://auto-stopped";
pub const LIVE_TRANSCRIPT: &str = "live://transcript";
pub const LIVE_TRANSLATION: &str = "live://translation";
pub const SESSIONS_CHANGED: &str = "sessions://changed";
pub const TRANSCRIBE_PROGRESS: &str = "transcribe://progress";
pub const TRANSCRIBE_DONE: &str = "transcribe://done";
pub const TRANSCRIBE_ERROR: &str = "transcribe://error";
pub const MODEL_PROGRESS: &str = "model://progress";
pub const MINUTES_DONE: &str = "minutes://done";
pub const PLAYER_POSITION: &str = "player://position";

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RecordingLevels {
    pub sys: f32,
    pub mic: f32,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RecordingStatePayload {
    pub recording: bool,
    pub duration: f64,
    pub session_id: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct NoInputWarning {
    pub sys: bool,
    pub mic: bool,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LiveTranscriptPayload {
    pub source: String,
    pub finalized: String,
    pub volatile: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LiveTranslationPayload {
    pub source: String,
    pub text: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TranscribeProgress {
    pub session_id: String,
    pub phase: String,
    pub fraction: Option<f64>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TranscribeDone {
    pub session_id: String,
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ModelProgress {
    pub model_id: String,
    pub fraction: f64,
    pub error: Option<String>,
    pub done: bool,
}

pub fn emit<P: Serialize + Clone>(app: &AppHandle, event: &str, payload: P) {
    if let Err(e) = app.emit(event, payload) {
        log::warn!("failed to emit {event}: {e}");
    }
}
